app/src/main/java/com/tavarus/artabletop/Controllers/AuthController.kt

```diff
@@ -0,0 +1,51 @@
+package com.tavarus.artabletop.Controllers
+
+import android.util.Log
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.auth.FirebaseUser
+import com.tavarus.artabletop.MainActivity
+import java.lang.Exception
+
+class AuthController() {
+    companion object {
+        var INSTANCE: AuthController = AuthController()
+    }
+
+    private var auth: FirebaseAuth = FirebaseAuth.getInstance()
+
+    fun logIn(email: String, password: String, activity: MainActivity, onSuccess: (FirebaseUser)->Unit, onFailure: (Exception?)->Unit ) {
+        auth.signInWithEmailAndPassword(email, password)
+            .addOnCompleteListener(activity) { task ->
+                if (task.isSuccessful) {
+                    onSuccess(auth.currentUser!!)
+                    activity.onLogin()
+                } else {
+                    // If sign in fails, display a message to the user.
+                    onFailure(task.exception)
+                }
+            }
+    }
+
+    fun signUp(email: String, password: String, activity: MainActivity, onSuccess: (FirebaseUser)->Unit, onFailure: (Exception?)->Unit ) {
+        auth.createUserWithEmailAndPassword(email, password)
+            .addOnCompleteListener(activity) { task ->
+                if (task.isSuccessful) {
+                    onSuccess(auth.currentUser!!)
+                    activity.onLogin()
```

Comment:
This inherently ties the `AuthController` to only be used within the context of `MainActivity`. It would be better to create an `AuthControllerCallback` interface that `MainActivity` implements, then require `AuthControllerCallback` as a parameter instead of `MainActivity`.

Reply:
That makes sense for code strength(? I'm blanking on the term), but I feel like `MainActivity` will be the class that handles this for the foreseeable future, and it wouldn't be too hard to update in the future if that use case comes up. I think it's worth the readability/smoothness now as a sacrifice for upgradeability later.